src/QsCompiler/Tests.Compiler/SyntaxTests.fs

```diff
@@ -17,7 +17,40 @@ open Xunit
 
 let private rawString = getStringContent (manyChars anyChar) |>> fst
 
-// Component parsers
+
+[<Fact>]
+let ``Reserved patterns`` () = 
+    [
+        ("_mySymbol" , true , Some "_mySymbol" , [])
+        ("mySymbol_" , true , Some "mySymbol_" , [])
+        ("my_symbol" , true , Some "my_symbol" , [])
+        ("my__symbol", true , Some "my__symbol", [Warning WarningCode.UseOfUnderscorePattern])
+        ("__mySymbol", true , Some "__mySymbol", [Warning WarningCode.UseOfUnderscorePattern])
+        ("mySymbol__", true , Some "mySymbol__", [Warning WarningCode.UseOfUnderscorePattern])
+        ("__my__sym" , true , Some "__my__sym" , [Warning WarningCode.UseOfUnderscorePattern])
+        ("my__sym__" , true , Some "my__sym__" , [Warning WarningCode.UseOfUnderscorePattern])
+        ("__mysym__" , true , None             , [Error ErrorCode.InvalidUseOfReservedKeyword])
+    ]
+    |> List.iter (testOne (symbolNameLike ErrorCode.InvalidIdentifierName))
+    [
+        ("a.b" , true , ([Some "a" ],Some "b"), [])
+        ("_a.b" , true , ([Some "_a" ],Some "b"), [])
+        ("a_.b" , true , ([Some "a_" ],Some "b"), [Warning WarningCode.UseOfUnderscorePattern])
```

Comment:
To clarify, are trailing single underscores reserved as well?

Reply:
A trailing single underscore at the end of a callable/type name is fine, but not at the end of a namespace name (since the full name of a callable would then have that underscore be followed by a dot).

Reply:
Got it, thanks for clarifying. Is there somewhere that would make sense document that, perhaps?

Reply:
Right now it's just a warning. We actually currently don't document in our docs repo what valid identifiers in Q# are but expect to do that when we open the language repo. Hopefully that should be in time for the change next month.